### PR TITLE
net: expose keepalive configuration on `TcpSocket`

### PR DESCRIPTION
**Pull request by hawkw**
Depends on #3145

## Motivation

Tokio v0.3 no longer exposes a way to configure TCP keepalive probes.
Many users need this functionality, so we want to re-expose keepalive
configuration via `TcpSocket`.

## Solution

This commit exposes the TCP keepalive configuration added to `mio`'s
`TcpSocket` in tokio-rs/mio#1385. This includes a method for enabling
keepalvie and checking if keepalive is enabled, a builder interface for
configuring keepalive parameters (keepalive time, the amount of time
before keepalive probes are sent, keepalive interval, the time between
each probe, and keepalive retries, the number of failed probes before a
connection is closed), and accessors for the values of keepalive
parameters. Due to Windows API limitations, `mio` has to provide a
builder for configuring keepalive parameters, since they must all be set
in one API call. Therefore, we re-export the keepalive builder struct.
If the `mio` API changes, we should be able to replace this with our own
type transparently, so it didn't seem necessary to wrap this --- if I'm
wrong, please let me know!

Fixes: #3082

### CODE REVIEW

File: tokio/src/net/tcp/socket.rs [hunk 1]
```diff
@@ -86,6 +87,11 @@ cfg_net! {
     }
 }
 
+cfg_net! {
+    #[allow(unreachable_pub)] // https://github.com/rust-lang/rust/issues/57411
+    pub use mio::net::TcpKeepalive;
```
Comment (carllerche) [inline tokio/src/net/tcp/socket.rs:92]: We are not exposing mio as part of Tokio's public API. If we do this, we will need to re-define the type.

File: tokio/src/net/mod.rs [hunk 2]
```diff
@@ -33,7 +33,7 @@ cfg_net! {
 
     pub mod tcp;
     pub use tcp::listener::TcpListener;
-    pub use tcp::socket::TcpSocket;
+    pub use tcp::socket::{TcpSocket, TcpKeepalive};
```
Comment (carllerche) [inline tokio/src/net/mod.rs:36]: We probably can keep this type in `tokio::net::tcp` and not re-export it.

File: tokio/src/net/tcp/socket.rs [hunk 3]
```diff
@@ -280,6 +286,255 @@ impl TcpSocket {
...
+    /// enabled on this socket.
+    ///
+    /// If `None`, then keepalive messages are disabled.
+    ///
+    /// This returns the value of `TCP_KEEPINTVL` on supported Unix operating
+    /// systems. On Windows, it is not possible to access the value of TCP
+    /// keepalive parameters after they have been set..
+    ///
+    /// Some platforms specify this value in seconds, so sub-second
+    /// specifications may be omitted.
```
Comment (bdonlan) [inline tokio/src/net/tcp/socket.rs:487]: Rounded down or up? What happens if this rounds to zero?

File: tokio/src/net/tcp/socket.rs (hunk 1 above)
Comment (bdonlan) [inline tokio/src/net/tcp/socket.rs:92]: This makes a mio type part of the public API, which is probably not ideal.

**Review by carllerche — CHANGES_REQUESTED** [review]
I'm hesitant to add this API at this time. This goes beyond what Tokio has done to date with regard to socket options. Until now, we have added socket option APIs that are essentially pass-through to the OS call. This attempts to provide a portable API that abstracts differences.

I also see a potential issue with the implementation in Mio where a single call to `set_keepalive_params` may result in multiple sub calls. If the first sub call succeeds, but the second fails, the socket state is inconsistent. All other socket option setters are atomic.

Perhaps it is best to leave such configuration to other crates or the user to call themselves?

**Comment by carllerche** [comment]
Closing as we decided in Discord to punt this.

**Comment by hawkw** [comment]
@carllerche wdyt about exposing _just_ keepalive time (time to the initial probe)? That, at least, can be configured consistently across all platforms.

**Comment by carllerche** [comment]
If it is consistent across platforms then it seems fine to me.

**Pull request closed without merging**